aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -6,6 +6,7 @@
 #include <ATen/native/TypeProperties.h>
 #include <ATen/MemoryOverlap.h>
 #include <ATen/native/Resize.h>
+#include <ATen/core/grad_mode.h>
```

Comment:
Using `AutoNonVariableTypeMode` here is a good thing (I think) but I don't think you need `grad_mode.h` for `AutoNonVariableTypeMode`. `GradMode` is a standalone thread-local flag, a temporary hack separate from the the dispatch hierarchy.  `AutoNonVariableTypeMode` controls the true autograd dispatch keys, which determine whether an op dispatches through `VariableType*` at all.  `GradMode` is a separate flag that determines if grad history recording takes place if you happen to route through `VariableType*`.

While working on the torchvision PR I beat my head through the [distinction between the two](https://github.com/pytorch/pytorch/issues/40736#issuecomment-652129302).  my comment is almost two months old, but given that the [TODO to unify GradMode and the autograd dispatch keys](https://github.com/pytorch/pytorch/blob/7f967c08b8ae47bd400df53caafa5061e9bc3bb4/aten/src/ATen/core/LegacyTypeDispatch.h#L39-L44) in `LegacyTypeDispatch.h` is still present, the fact that autograd dispatch selection [still leaves DispatchKey::Autograd always-on-by-default](https://github.com/pytorch/pytorch/blob/7f967c08b8ae47bd400df53caafa5061e9bc3bb4/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L18), and the fact that GradMode is still present in Pytorch at all, my comment is not obsolete.  @ezyang would know more.